src/metabase/search/spec.clj

```diff
@@ -0,0 +1,255 @@
+(ns metabase.search.spec
+  (:require
+   [clojure.set :as set]
+   [clojure.string :as str]
+   [clojure.walk :as walk]
+   [malli.core :as mc]
+   [malli.error :as me]
+   [metabase.config :as config]
+   [metabase.util :as u]
+   [toucan2.core :as t2]))
+
+(def ^:private SearchModel
+  [:enum "dashboard" "table" "dataset" "segment" "collection" "database" "action" "indexed-entity" "metric" "card"])
+
+(def ^:private AttrValue
+  "Key must be present, to show it's been explicitly considered.
+
+  - false: not present
+  - true: given by a column with the same name (snake case)
+  - keyword: given by the corresponding column
+  - vector: calculated by the given expression
+  - map: a sub-select"
+  [:union :boolean :keyword vector? :map])
+
+(def ^:private explicit-attrs
+  "These attributes must be explicitly defined, omitting them could be a source of bugs."
+  [:archived
+   :collection-id
+   :database-id
+   :table-id])
+
+(def ^:private optional-attrs
+  "These attributes may be omitted (for now) in the interest of brevity in the definitions."
+  [:id ; unlike the others, this is implictly true
```

Comment:
I think it'll benefit us if we document the "not so obvious" fields a bit.

Like `pinned`, I'd document where it came from and what we use it for.

Reply:
Yeah this is a good point, it'd be good to know how all the fields are being used, only `:search-terms` are unambiguous really.

Attrs are all saved as database columns, but some are used for filtering, some for scoring.

Some of the render fields are also used for scoring.

This will get a bit better when we rewrite scoring to happen fully in the database I think, and then we can do a final polish PR after that making the docstrings and comments shine 🌻